### ISSUE TITLE
Capybara and microservices (concurrent re-entering server request)

### ISSUE THREAD

**Issue by rosskevin** [teamcapybara/capybara#1934]
## Meta
Capybara Version: 2.15.2
Driver Information (and browser if relevant): selenium 3.6.0

```ruby
::Capybara.server = :puma
::Capybara.server_host = '0.0.0.0'
::Capybara.server_port = 4100
::Capybara.register_driver :chrome_headless

RSpec.describe Agreement, :feature, js: true, driver: :chrome_headless do
...
end
```

I have both rspec and capybara integrated with `:js` selenium tests that have been working for quite some time.  We are just now switching to more microservices, so new complexity has been added to the testing workflow.

## Expected Behavior
Reliably run capybara server to receive and process a concurrent/re-entering request at the running server url e.g. `http://foo.com.0.0.0.0.nip.io:4100/bar`

## Actual Behavior
A concurrent re-entering server request is not served. 

NOTE: occasionally during debugging play/pause I can get it to respond to a browser, so I know network setup is correct, but it isn't reliable (high latency) and my rspec tests fail quickly.

## Research
For reasons listed by @twalpole in https://github.com/teamcapybara/capybara/issues/1403#issuecomment-56321208, I understand why this is not the default.  It references this commit 385a7507f6525d9b2d1e23bef0bb2e6fe5ad0c97 from #1400.  But, I see this code is now not present after https://github.com/teamcapybara/capybara/commit/a4fc9d96f705e9ac4732c5fa05ab79d947743599#diff-854ad13f608ddda187f483f7df075752

So it seems that my capybara server should be serving concurrent requests?

## Steps to reproduce
![https___mermaidjs_github_io_mermaid-live-editor___edit_c2vxdwvuy2veawfncmftcnbhcnrpy2lwyw50iejyb3dzzxikcgfydgljaxbhbnqgqvbjcnbhcnrpy2lwyw50ierccnbhcnrpy2lwyw50ifnpzgvraxfkb2ikcgfydgljaxbhbnqgumvkaxmkcgfydgljaxbhbnqguhvwcgv0cm9ucgpsb29wifvzz](https://user-images.githubusercontent.com/136564/32382166-96d7df12-c082-11e7-8bc6-632cdf1a0960.png)

Note that in this diagram above when run via rspec, I am running _on the same thread_ as an inline `perform_one` execution of the job.  I have also validated that the server is successfully running with an initial `visit /` in the rspec test.

## TL;DR
I'm not implying this is a bug, but I need some guidance on how I can get/allow capybara server to respond to this concurrent re-entering request.  I'm open to suggestions and ready to try anything!

**Comment by rosskevin** [comment]
Please note that I have also tried with:
```ruby
config.cache_classes = true
config.eager_load = true
```

**Comment by twalpole** [comment]
Capybara isn't doing anything to prevent concurrent requests and as you stated this isn't a bug in Capybara.  

I assume `::Capybara.register_driver :chrome_headless` has more to it than that (like a block being passed)???

"I am running on the same thread as an inline perform_one execution of the job" - Not really sure what you mean by this, it's not possible to run two things in the same thread simultaneously.  

Also, you don't mention what version of Rails you're using, whether you're using database_cleaner, how you're starting up your micro service during testing. 

I'm going to close this since it's not a Capybara issue, if you want to keep adding more info (probably including some test code that's failing) I will take a glance as it becomes clearer exactly what you're doing or you can post it any of mailing list/gitter/stackoverflow if you want to get more eyes on it

**Comment by rosskevin** [comment]
I updated versions, and yes using database cleaner.  Again, all is well (and has been for years) for a very large suite of rspec and cucumber tests.  This issue is only related to the beginning of moving to containerized microservices (pdf rendering is one).

> I assume ::Capybara.register_driver :chrome_headless has more to it than that (like a block being passed)???

Yes, and I omitted it because it it successfully up and running.  I just wanted to note that I have connected everything and am consistently using one named driver.

> "I am running on the same thread as an inline perform_one execution of the job" - Not really sure what you mean by this, it's not possible to run two things in the same thread simultaneously.

- Rspec is executing on a thread
- Capybara server is executing on one or more threads?
- [In testing, the sidekiq job is executing on the main/rspec thread. This is known as inline mode](https://github.com/mperham/sidekiq/wiki/Testing), so a separate worker thread/process is not executing concurrently, allows for easier testing of jobs.

## Boiled down to the essentials
The main thread (rspec/sidekiq job) is trying an http request, which synchronously comes back to the Capybara server, so the main (rspec/sidekiq) thread is blocked until the Capybara server responds.  I assume by your comment that this means regardless if the rspec thread is blocking, Capybara should serve many requests?

Put another way, if start an rspec test and `sleep 6000000000`, Capybara should be serving web requests without blocking?

If the debugger has the main thread paused, should Capybara still serve?

**Comment by rosskevin** [comment]
So, a little proof that Capybara server is running/serving:

## Rspec test

```ruby
    it 'sleeps for 50 years' do
      visit '/'
      expect(page).to have_content('Landing Page')
      puts 'Winter is coming...'
      sleep 50 * 31556952
    end
```

Open a browser to `http://foo.com.0.0.0.0.nip.io:4100` and see it serve.

So, not sure why it isn't serving to my job yet, but the above proves Capybara is working/serving.

**Comment by twalpole** [comment]
Depending on the debugger it may be pausing all the threads -- and IIRC there is an issue with byebug on MacOS where readline causes any other threads to block.  

Inline mode for sidekiq just runs the job immediately where it's queued, so it would actually be running in the thread running the app (Capybara server thread)

If you're not using the debugger it sounds like something in your app/rails is preventing multiple requests from being served simultaneously by multiple threads.  In Rails 5.1 when in test mode all threads share the same database connection (via a mutex - and why database_cleaner is generally not needed anymore in rails 5.1) so it's possible the second request is blocked when it attempts to access the database - does the second request show anything in the test.log?

**Comment by twalpole** [comment]
http://guides.rubyonrails.org/5_1_release_notes.html - Section 3

**Comment by twalpole** [comment]
It's also possible that however you are making the second request is blocking the GIL and preventing that second request from being processed in a new thread.
